networks/2b_community_detection.py

```diff
@@ -0,0 +1,202 @@
+# User params
+
+# Choices = ["fastgreedy", "walktrap", "louvain", "infomap"]
+method = "infomap"
+
+# Params for different methods to adjust
+# length of random walk to perform for walktrap
+# Short random walks tend to stay in the same community
+nsteps = 10
+
+# Number of trials to attempt to partition the network for infomap
+ntrials = 20
+
+# TO DO
+# Think about running these methods across multiple seeds and taking a consensus
+
+# +
+# Load correlation matrix
+pao1_pearson_mat_filename = paths.PAO1_CORR_LOG_SPELL
+pa14_pearson_mat_filename = paths.PA14_CORR_LOG_SPELL
+
+# Take abs of correlation scores
+# In this case we care about the strength and not the direction
+pao1_corr = pd.read_csv(
+    pao1_pearson_mat_filename, sep="\t", index_col=0, header=0
+).abs()
+pa14_corr = pd.read_csv(
+    pa14_pearson_mat_filename, sep="\t", index_col=0, header=0
+).abs()
+# -
+
+pao1_corr.head()
+
+# +
+# Format correlation matrix into graph (i.e. dataframe with edge weight per pair of genes)
+# The dataframe should have columns: from, to, weight
+pao1_corr_graph = pao1_corr.stack().reset_index()
+pao1_corr_graph.columns = ["from", "to", "weight"]
+
+pa14_corr_graph = pa14_corr.stack().reset_index()
+pa14_corr_graph.columns = ["from", "to", "weight"]
+# -
+
+# Drop duplicate rows since correlation matrix is symmetric
+pao1_corr_graph = pao1_corr_graph.drop_duplicates()
+pa14_corr_graph = pa14_corr_graph.drop_duplicates()
+
+# Drop gene loops
+# Note 'query' not working for some reason
+pao1_corr_graph = pao1_corr_graph[pao1_corr_graph["from"] != pao1_corr_graph["to"]]
+pa14_corr_graph = pa14_corr_graph[pa14_corr_graph["from"] != pa14_corr_graph["to"]]
+
+print(pao1_corr_graph.shape)
+pao1_corr_graph.head()
+
+# Make into a graph object
+pao1_G = ig.Graph.TupleList(pao1_corr_graph.values, weights=True, directed=False)
+pa14_G = ig.Graph.TupleList(pa14_corr_graph.values, weights=True, directed=False)
+
+# make sure vertex/edge properties exist
+print(pao1_G.es["weight"][:5])
+
+# ## Community detection
+
+# ### Fast-greedy
+# This algorithm starts from a completely unclustered set of nodes and iteratively adds communities such that the modularity (score maximizing within edges and minimizing between edges) is maximized until no additional improvement can be made.
+#
+# Note: Looks like fast-greedy requires a simple graph (i.e. no multiple edges per node), so we use [simplify](https://igraph.org/python/doc/api/igraph._igraph.GraphBase.html#simplify) to combine edges
+#
+# Returns VertextDendrogram
+
+# %%time
+if method == "fastgreedy":
+    # Simplify graph to remove multiple edges and loops
+    if not pao1_G.is_simple():
+        pao1_G.simplify()
+    if not pa14_G.is_simple():
+        pa14_G.simplify()
+
+    assert pao1_G.is_simple()
+    assert pa14_G.is_simple()
+
+    # Detection method
+    pao1_partition = pao1_G.community_fastgreedy(weights=pao1_G.es["weight"])
+    pa14_partition = pa14_G.community_fastgreedy(weights=pa14_G.es["weight"])
```

Comment:
Is the partition for these methods a 1d array with integers like the _labels property of sklearn's clustering algorithms?

Reply:
So the partition is either a `VertexClustering` or `VertexDendrogram` object that is essentially a dictionary mapping module ids or labels to vertex ids. This information can be combined with the graph which contains attributes for vertex and edges which then have vertex attributes like "name" and edge attributes like "weight". When we combine this information we can get the name of the vertex and which module id/label it maps to as well as the degree